src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,225 @@
+            {
+                typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                var paramReference = (IParameterReferenceOperation)context.Operation;
+
+                switch (paramReference.Parent.Kind)
+                {
+                    case OperationKind.Invocation:
+                        AddDeclaringTypeSymbol(((IInvocationOperation)paramReference.Parent).TargetMethod);
+                        break;
+
+                    case OperationKind.PropertyReference:
+                    case OperationKind.EventReference:
+                    case OperationKind.FieldReference:
+                    case OperationKind.MethodReference:
+                        AddDeclaringTypeSymbol(((IMemberReferenceOperation)paramReference.Parent).Member);
+                        break;
+
+                    case OperationKind.Argument:
+                        AddType(paramReference.Type);
+                        break;
+
+                    case OperationKind.Conversion:
+                        AddType(paramReference.Parent.Type);
+                        break;
+
+                    default:
+                        break; // Do nothing
+                }
+
+                // Local functions
+                void AddType(ITypeSymbol typeSymbol)
+                {
+                    if (typesUsedPerParameter.TryGetValue(paramReference.Parameter, out var typePool))
+                    {
+                        typePool.Add(typeSymbol);
+                    }
+                }
+
+                void AddDeclaringTypeSymbol(ISymbol symbol)
+                    => AddType(FindDeclaringType(symbol));
+            }, OperationKind.ParameterReference);
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                try
+                {
+                    var overloads = methodSymbol.GetOverloads();
+
+                    foreach (var (parameter, usedTypesPool) in typesUsedPerParameter)
+                    {
+                        var usedTypes = usedTypesPool.ToImmutableArray();
+                        // If type is unused, we ignore it. There is another rule to detect
+                        // unused parameters.
+
+                        if (usedTypes.Length == 1)
+                        {
+                            ReportIfTypesAreDifferent(parameter, usedTypes[0], overloads);
+                        }
+                        else if (usedTypes.Length > 1)
+                        {
+                            ReportIfTypesAreDifferent(parameter, FindMostGenericType(parameter.Type, usedTypes), overloads);
+                        }
+                    }
+                }
+                finally
+                {
+                    typesUsedPerParameter.Free(context.CancellationToken);
```

Comment:
Does this also free the PooledConcurrentSet values in this dictionary?

Reply:
I don't know. `Free()` is calling `Clear` which I assume calls `Dispose` which would call the `Free` but I am not sure so I will add a loop.

Reply:
I have added a loop.